omod/src/main/java/org/openmrs/module/coreapps/page/controller/MergeVisitsPageController.java

```diff
@@ -52,27 +64,54 @@ public Object controller(@RequestParam("patientId") Patient patient,
 
     public String post(@RequestParam("patientId") Patient patient,
                        @RequestParam("mergeVisits") List<Integer> mergeVisits,
-					   @RequestParam(value = "returnUrl", required = false) String returnUrl,
+                       @RequestParam(value = "returnUrl", required = false) String returnUrl,
                        @SpringBean AdtService service,
                        UiUtils ui,
-                       HttpServletRequest request, PageModel model){
+                       HttpServletRequest request, PageModel model) {
 
+        List<String> mergedVisitIds = new ArrayList<String>();
         if (patient.isVoided() || patient.isPersonVoided()) {
             return "redirect:" + ui.pageLink("coreapps", "patientdashboard/deletedPatient",
-					SimpleObject.create("patientId", patient.getId().toString() ));
+                    SimpleObject.create("patientId", patient.getId().toString()));
         }
 
-        if (mergeVisits!=null && mergeVisits.size() > 0 ){
+        if (mergeVisits != null && mergeVisits.size() > 0) {
             Visit mergedVisit = service.mergeConsecutiveVisits(mergeVisits, patient);
-            if (mergedVisit != null){
+            if (mergedVisit != null) {
                 request.getSession().setAttribute(AppUiConstants.SESSION_ATTRIBUTE_INFO_MESSAGE, ui.message("coreapps.task.mergeVisits.success"));
                 request.getSession().setAttribute(AppUiConstants.SESSION_ATTRIBUTE_TOAST_MESSAGE, "true");
-            }else{
+            } else {
                 request.getSession().setAttribute("emr.errorMessage", ui.message("coreapps.task.mergeVisits.error"));
                 request.getSession().setAttribute(AppUiConstants.SESSION_ATTRIBUTE_TOAST_MESSAGE, "true");
             }
+
+            //Pass by all visits except the first one.
+            for (int i = 1; i < mergeVisits.size(); i++) {
+                mergedVisitIds.add(mergeVisits.get(i).toString());
+            }
         }
+        return "redirect:" + ui.pageLink("coreapps", "mergeVisits", SimpleObject.create("patientId", patient.getId(), "returnUrl", returnUrl, "mergedVisitIds", mergedVisitIds));
+    }
+
 
-		return "redirect:" + ui.pageLink("coreapps", "mergeVisits", SimpleObject.create("patientId", patient.getId(), "returnUrl", returnUrl));
+    /**
+     * If some of the merged visit IDs are on the return url, it scraps the visitId param altogether off the URL
+     * so it doesn't return to a visit that no longer exists.
+     */
+    protected String removeVisitIdFromURLIfVisitWasMerged(String returnUrl, List<String> mergedVisitsIDs) {
+        //Pattern to get the visitID from url
+        Pattern pat = Pattern.compile("(?<=visitId=)[^&]+");
+        Matcher urlVisitId = pat.matcher(returnUrl);
+        //If find some visitID on url.
+        if (urlVisitId.find()) {
+            for (String mergedVisitId : mergedVisitsIDs.get(0).split(",")) {
```

Comment:
Is there a possibility of `mergedVisitsIDs` being empty?

Reply:
No, there is a validation for that. I am only able to merge if I select at least 2 visits.